getzig: Add script to download zig toolchain

camh- opened the pull request:
Add a script to download zig releases, similar to `getgo`. It installs
it in `/opt/zig/<version>` and symlinks `/opt/zig/{stable,master}` to it
(depending on whether you have opted to download the latest stable or
master release.

camh- committed:
getzig: Add script to download zig toolchain

Add a script to download zig releases, similar to `getgo`. It installs
it in `/opt/zig/<version>` and symlinks `/opt/zig/{stable,master}` to it
(depending on whether you have opted to download the latest stable or
master release.

camh- commented on getzig:
```diff
@@ -0,0 +1,115 @@
+
+case $(uname -m) in
+x86_64) ARCH=x86_64 ;;
+aarch64) ARCH=aarch64 ;;
+arm64) ARCH=aarch64 ;;
```
Hmmm, I should combine these:
```
aarch64|arm64) ARCH=aarch64 ;;
```

juliaogris replied:
sounds reasonable.

camh- replied:
done

camh- commented on getzig:
```diff
@@ -0,0 +1,115 @@
+
+# debug sends stdin to stdout, and if debug mode is on, also sends stdin to
+# stderr.
+debug() {
```
I think I should rename this to `dumpstdin` or something

juliaogris replied:
debug makes sense to me, lines up with GETZIG_DEBUG

camh- replied:
ok. i left it as debug

camh- commented on getzig:
```diff
@@ -0,0 +1,115 @@
+	fi
+}
+#-----------------------------------------------------------------------------
+[[ "$(caller)" != 0\ * ]] || main "$@"
```
I should change this to my new construct: 
```
# Only run main if executed as a script and not sourced.
if [[ "${BASH_SOURCE[0]}" == "$0" ]]; then main "$@"; fi
```

juliaogris replied:
sounds good.

camh- replied:
done

juliaogris reviewed (commented):
I think you should add `shfmt` to make / CI.
Maybe also a note for deps on newer `bash` / `jq` / `curl` ?
With bad input I get: 
```
./getzig: line 40: info["${key}"]: bad array subscript`
```

juliaogris commented on getzig:
```diff
@@ -0,0 +1,115 @@
+		;;
+	esac
+
+	wetrun curl -sL "https://ziglang.org/download/index.json" |
```
🍭

juliaogris replied:
to me wetrun is a confusing term. alwaysrun or forcerun would make more sense to me.
(you choose dry, so i would have expected that you chose wet)...
feel free to ignore if this is a common pattern / doesn't resonate.

camh- replied:
ok. i changed it to `alwaysrun`

camh- commented:
> Maybe also a note for deps on newer `bash` / `jq` / `curl` ?

I've added a check for all these deps so it give a useful error.

> With bad input I get: [...]

Yeah, that was a bug. I've fixed it I think. The `set -e` was not taking effect because of where `get_latest_info` was run. I've moved that so it is returns an error, the script will error instead of continuing and operating on bad data.

> I think you should add shfmt to make / CI.

Ok. But not this PR. I would need to reformat getgo first/concurrently.

camh- commented:
I moved the os/arch stuff into `get_latest_info` since that's the only place they were used. Now everything is in functions as it should be.

juliaogris reviewed (approved):
Lgtm, but I only re-reviewed on phone. If you want I try running it again on my Mac tomorrow.
🟩
